---
layout: post
title:  Ian Korf Unit 1 - Lab 2  
date:   2014-04-08 00:14:21
categories: 
---

## Installing BLASTP ##

The `blastp` program aligns proteins to each other in a manner similar
to `water` but `blastp` is much faster. Our first task is to download
and install `blastp` and other tools in the BLAST suite. There are two
versions of BLAST, NCBI-BLAST and AB-BLAST. We are going to use
AB-BLAST. It's not as common as NCBI-BLAST, but it is superior (sort of
like how Macs are superior to PCs). Start by going to the [AB-BLAST
download site](http://blast.advbiocomp.com/licensing) to fill out a
license form. Make sure you scroll all the way to the bottom in the text
box.

Uncompress the tarball into a new directory where BLAST will live.

    mkdir ~/BLAST
    cd ~/BLAST
    mv ~/Downloads/blast3.pe.linux26-x64.tar.gz .
    tar -zxf blast3.pe.linux26-x64.tar.gz

We also need to tell our shell that we have some new programs we want to
make available on the command line. You have to edit your `~/.profile`
file to update the PATH variable to include `~/BLAST`. Include the
following line at the end of the file.

    PATH="$HOME:/BLAST:$PATH"

This tells the shell that the paths to your applications should also
include `~/BLAST`. The colon character separates the various
directories. If you want to see all the locations in your `PATH`, do the
following.

    printenv PATH

From now on, every time you open a new terminal, your PATH will include
the various BLAST programs. To update your current shell, you must do
the following:

    source ~/.profile

Test `blastp` to make sure it works.

    blastp

That should produce a frighteningly long usage statment. If you got an
error saying `command not found` you either made an error in your
`.profile` or you didn't `source` it.


## Part 1: Orthologs ##

In the last lab, you learned that

* There is always a best local alignment, even in random sequences
* The distribution of random alignment scores is not normal
* The random background depends on
    * The lengths of the sequences
    * The composition of the sequences
    * The scoring parameters (matrix, gaps)
* The significance of a score is difficult to assess

We previously called the best match to B0213.10 in the A. thalian and D.
melanogaster genomes the ortholog. but the ortholog is operationally
defined as the best reciprocal match between proteomes. That is, after
finding the best match to the fly genome, one must take the fly protein
and search it against the worm proteome to determine if it finds
B0213.10. If if it does, the proteins are orthologous.

Let's do something ambitious. Let's find ALL the orthologous proteins
between worm and fly. How long will it take to search every protein
against every other protein? You made this calculation in Lab 1 with the
`water` program and it was outrageously long. Let's start with the same
task but using `blastp`. To begin, let's organize our thoughts and files
in a new directory.

    mkdir ~/Lab2
    cd ~/Lab2

Before you can use `blastp` you need to format the BLAST database. A
FASTA file can be turned into a BLAST database with `xdformat`. Let's
make an alias to a FASTA proteom, format it as a BLAST database, and
then observe the new files.

    ln -s ~/Data/Species/D.melanogaster/protein.fa ./flypep
    xdformat -p flypep
    ls -lrt

Now let's try aligning our previous favorite protein (`B0213.10`) to the
fly proteome. Make an alias here and call it `P450`. We know from Lab 1
that comparing a single C. elegans protein to a entire proteome can take
a couple minutes, and that if you attempted to align two whole
proteomes, it would take weeks. Let's align P450 to the entire D.
melanogaster proteome with BLAST and see how long it takes to run.

    time blastp flypep P450 > default.blast

That was fast! `blastp` is clearly much faster than `water`. Not only
that, if you look at the files, you will see that a BLAST report gives
you some statistics about the search. The E-value is sort of similar to
what you would get if you did some shuffling experiments and determined
how different the alignment was from the random background.

There are a lot of parameters that contol the speed of BLAST. The most
important of these are the seeding parameters. The default `blastp`
search uses `W=3 T=11`. We can change the seeding parameters quite a
bit. We will set `T=999` so that there is no **neighborhood** aound each
word match. In other words, only exact matches will produce a seed.
Let's vary the word match from 1 to 6.

    time blastp flypep P450 T=999 W=1 > w1.blast
    time blastp flypep P450 T=999 W=2 > w2.blast
    time blastp flypep P450 T=999 W=3 > w3.blast
    time blastp flypep P450 T=999 W=4 > w4.blast
    time blastp flypep P450 T=999 W=5 > w5.blast
    time blastp flypep P450 T=999 W=6 > w6.blast

Now calculate how long it will take to search a proteome against another
proteome using the parameters above and add these to your notebook.

---------------------------------------------------------------------------

Time to set up the searches. First thing will be to make a `wormpep`
database (make an alias and then xdformat it). Then you can do the
search as follows.

    blastp wormpep flypep mformat=3 hspmax=1 E=1e-15 W=5 warnings

A few command line switches were added. `mformat=2` made the output
tabular. `hspmax=1` made it so that only the best alignment was shown
(there can be more than one unlike `water`). `warnings` turns off some
warning messages. You have to do the search in the other direction too.
You can start that process now. Since the processes are running on
different CPUs, they will not interfere with each other (much).

    blastp flypep wormpep mformat=3 hspmax=1 E=1e-15 W=5 warnings

While you're waiting for the jobs to complete, you might want to tackle
the 2nd half of this lab farther below.

-------------------------------------------------------------------------

Once the jobs are complete use `less` to examine the outputs of each
(actually, you can do this even before the jobs are complete). Columns
1, 2, and 3 contain the names of the proteins and the E-value of the
match. If it makes it easier, you can `cut` these, but you might want to
look at other columns of the file.

    cut -f 1-3 __whatever_you_named_your_file__

Use the `less` search feature (forward slash key) to find proteins in
one file that match proteins in the other file. Find some putative
orthologs. Answer the following questions in your notebook.

* What difficulties are there in finding orthologs?
* Which BLAST parameters might you change if you did this again?
* What additional information would you like to have?
* What logic would you use to find orthologs automatically?


## Part 2: P450 Paralogs ##

In this part, we are going to look for the **paralogs** of the B0213.10
gene. That is, we are going to search the B0213.10 protein against the
C. elegans proteome to look for sequences that arose by duplication
within the same genome. A gene may have from 0 to many paralogs. A
simple, but flawed way to determine which alignments represent paralogs
is to use an E-value cutoff. For example, anything with an E-value less
than 1e-10 is a paralog.

    blastp wormpep P450 E=1e-10 > 1e-10.blast

Inspect the BLAST report with `less`. Do all the sequences look like
they are highly related? You may want to change the E-value cutoff to
something higher or lower than 1e-10.

### Notebook Questions ###

* What parameters did you choose for the search?
* How did you choose the parameters? What trade-offs were considered?
* What other properties would be useful for finding paralogs?
* How would you identify paralogs using BLAST?




